src/include/server/mir/shell/decoration.h

```diff
@@ -20,6 +20,122 @@
 #include "mir/geometry/forward.h"
 #include "mir_toolkit/common.h"
 
+#include "mir/scene/null_surface_observer.h"
+#include "mir/scene/surface.h"
+#include <memory>
+#include <mir/geometry/size.h>
+#include <mir/geometry/forward.h>
+#include <mir/events/event.h>
+
+#include <string>
+
+namespace mir
+{
+namespace shell
+{
+namespace decoration
+{
+/// Draws window decorations and provides basic move, resize, close, etc functionality
+class Decoration
+{
+public:
+    struct InputObserver : public mir::scene::NullSurfaceObserver
+    {
+        InputObserver(Decoration* decoration) :
+            decoration{decoration}
+        {
+        }
+
+        /// Overrides from NullSurfaceObserver
+        /// @{
+        void input_consumed(mir::scene::Surface const*, std::shared_ptr<MirEvent const> const& event) override
+        {
+            if (mir_event_get_type(event.get()) != mir_event_type_input)
+                return;
+
+            decoration->handle_input_event(event);
+        }
+        /// @}
+
+        Decoration* const decoration;
+    };
+
+    struct WindowSurfaceObserver : public mir::scene::NullSurfaceObserver
+    {
+    public:
+        WindowSurfaceObserver(Decoration* const& decoration) :
+            decoration{decoration}
+        {
+        }
+
+        /// Overrides from NullSurfaceObserver
+        /// @{
+        void attrib_changed(mir::scene::Surface const* window_surface, MirWindowAttrib attrib, int value) override
+        {
+            decoration->attrib_changed(window_surface, attrib, value);
+        }
+
+        void window_resized_to(mir::scene::Surface const* window_surface, geometry::Size const& window_size) override
+        {
+            decoration->window_resized_to(window_surface, window_size);
+        }
+
+        void renamed(mir::scene::Surface const* window_surface, std::string const& name) override
+        {
+            decoration->window_renamed(window_surface, name);
+        }
+        /// @}
+
+        Decoration* const decoration;
+    };
+
+    Decoration() = default; // For tests
+
+    Decoration(std::shared_ptr<mir::scene::Surface> decoration_surface, std::shared_ptr<mir::scene::Surface> window_surface) :
+        decoration_surface{decoration_surface},
+        window_surface{window_surface},
+        input_observer{std::make_shared<InputObserver>(this)},
+        window_observer{std::make_shared<WindowSurfaceObserver>(this)}
+    {
+        decoration_surface->register_interest(input_observer);
+        window_surface->register_interest(window_observer);
+    }
+
+    virtual ~Decoration()
+    {
+        if(decoration_surface)
+            decoration_surface->unregister_interest(*input_observer);
+
+        if(window_surface)
+            window_surface->unregister_interest(*window_observer);
+    }
+
+    // Called whenever the visible aspect of the window changes.
+    // Examples: Scale, title, focus, visibility, size, etc...
+    virtual void redraw() = 0;
```

Comment:
That sounds like it doesn't belong in the interface. Surely it is up to the implementation whether it uses a common `redraw()` method to process all these changes?

Alternatively, if this is the only correct way to implement this, then there should not be separate `set_scale()`, etc methods as well.

Reply:
There is no "right" way, really. Just different ways of doing the same thing. This is the most "easy to follow" interface I was able to extract from the code we currently have implemented. Additionally, I had in mind that a decoration author might want to redraw on every frame (for animations and such), so having a common redraw method is not a bad idea in my opionion.

`set_scale` and friends exist to allow decoration authors to be able to handle state changes, they could invoke `redraw` immediately inside of them, or just update the state and redraw at some other point.

Reply:
You're right, I originally meant to leave it to allow for us to maybe have fancier decoratios that update independently of input, but that's too far from now.